decomposition.tex

```diff
@@ -0,0 +1,378 @@
+    // requirements for move assignment
+    { a = move(b) } -> Same<T&>;
+    // requirements for move (de-) allocation
+    { new T{move(a)} } -> Same<T*>;
+    delete new T{move(a)};
+
+    // general object requirements that are
+    // here because Movable is the base of the
+    // concept hierarchy:
+    // addressability
+    { &t } -> Same<T*>;
+    // destruction
+    { t.~T() } noexcept;
+  };
+}
+
+template <class T>
+concept bool Copyable() {
+  return Movable<T>() && // refines Movable with
+    requires (T a, T b) {
+      // copy construction
+      T{a};
+      // copy assignment
+      { a = b } -> Same<T&>;
+      // copy (de-) allocation
+      { new T{a} } -> Same<T*>;
+      delete new T{a};
+    };
+}
+
+template <class T>
+concept bool Semiregular() {
+  return Copyable<T>() && // refines Copyable with:
+    requires {
+      // default construction
+      T{};
+      // default allocation
+      { new T{} } -> Same<T*>;
+      // array (de-) allocation
+      delete[] new T[1];
+    };
+}
+\end{codeblock}
+The \tcode{Copyable} and \tcode{Movable} concepts are necessary to describe both object types
+that are modeled by the standard library, and requirements on user types with which the standard
+library interacts. Simply changing those requirements to \tcode{Semiregular} would cause too
+much breakage of existing code, and result in inefficiencies when the old code is updated to meet
+the requirements of the stricter concept. This would violate the general ``Don't pay for what you
+don't use'' principle of \Cpp.
+
+Looking beyond the algorithms and iterators that were the focus of N3351 - and N4382, for that
+matter - the standard library requires concepts to describe:
+\begin{itemize}
+\item object types that are copy/move constructible but without assignment requirements
+\item object types that are constructible from a specific set of argument types
+\item object types that are only destructible (although the applicability of such a concept is
+narrow, it provides a convenient basis for the other object concepts)
+\end{itemize}
+We introduce concepts named \tcode{CopyConstructible}, \tcode{MoveConstructible},
+\tcode{Constructible}, \tcode{DefaultConstructible}, and \tcode{Destructible} to match these usages
+in the library, formed by further decomposition of \tcode{Copyable} and \tcode{Movable}. Although
+the Palo Alto Report warns against successive decomposition:
+\begin{quote}
+We recognize that these concepts could be further decomposed into smaller and smaller units.
+For example, we could easily see factoring out shared ``object requirements'' for the
+\tcode{Copyable}, \tcode{Movable}, and \tcode{Function} concepts. However, each subsequent
+refactoring yields concepts with less coherent meaning (if any).
+\end{quote}
+the proverbial ship has already sailed; the standard library uses these concepts in its design
+now. We simply provide convenient names by which to refer to them, and a coherent hierarchy
```

Comment:
One could argue that this was a mistake that should be remedied. At least, I am curious to know how disruptive it is to the Standard Library to try to make do only with Movable, Copyable, and Semiregular.